metrics/web/docs/features/deep_links/01_deep_linking.md

```diff
@@ -399,3 +412,153 @@ class MetricsRouteInformationParser {
+
+- Updating deep links in response to the UI state changes:
+  ![Saving deep links diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/deep_links_design_improvements/metrics/web/docs/features/deep_links/diagrams/updating_deep_links_sequence_diagram.puml)
+
+#### Internal app navigation
+This section describes the changes we should introduce to improve the overall navigation experience in the Metrics Web application.
+
+##### Back Button navigation
+Currently, the back button that is displayed on the `ProjectGroupPage` and on the `DebugMenuPage` navigates directly to the `DashboardPage`. However, this behavior can be unexpected to the user. Consider the following cases that may be confusing for the user:
+- The user navigates to the `ProjectGroupPage` from the `DebugMenuPage`, and presses the `Back Button`. The application navigates the user to the `DashboardPage` instead of the `DebugMenuPage`;
+- The user applies some filtering or sorting parameters on the `DashboardPage`, navigates to the `ProjectGroupPage`, and presses the `Back Button`. The application navigates the user to the `DashboardPage`, however, does not restore the page parameters.
+
+To improve the navigation experience here, we should introduce a new method to the `NavigationNotifier` - `canPop()`, that returns `true` if the current page can be popped.
+
+Now, when the `back button` is pressed, we should perform the following:
+1. Pop the current page if the `NavigationNotifier.canPop` returns `true`;
+2. If there is no underlying page, push and replace the current page with the `DashboardPage`.
+
+Using such an approach allows users to navigate directly to the previous page if it exists, and to restore the applied `PageParametersModel` to the previous page if any.
+
+##### Deep linking with authorization
+Currently, the application pushes the `DashboardPage` when the user logs in, however, it would be much more user-friendly if the application redirects the user to the requested resource.
+
+To improve this aspect of the navigation, we should modify the existing redirecting behaviour:
+1. Use the `_redirectRoute` field to stores the redirect `RouteConfiguration` that requires authorization.
+2. When the application pushes a new route that requires authorization, and the user is unauthenticated, set the `_redirectRoute`. If the route does not require authorization, clear the `_redirectRoute`.
+3. When the user logs in, redirect the user to the `_redirectRoute` and clear the `_redirectRoute`.
+
+To implement such a behaviour, we should split the `.handleAuthenticationUpdates()` method of the `NavigationNotifier` into two different methods: `.handleLogOut()` and `.handleLogIn()`.
```

Comment:
🔖 
What if we name these methods as below? It seems like they actually handle the event (not log-in but logged-in):
```suggestion
To implement such a behaviour, we should split the `.handleAuthenticationUpdates()` method of the `NavigationNotifier` into two different methods: `.handleLoggedOut()` and `.handleLoggedIn()`.
```